src/invocation/ClusterService.ts

```diff
@@ -14,317 +14,283 @@
  * limitations under the License.
  */
 
-import {ClientConnection} from './ClientConnection';
+import {ClientConnection} from '../network/ClientConnection';
 import * as Promise from 'bluebird';
-import {ClientAddMembershipListenerCodec} from '../codec/ClientAddMembershipListenerCodec';
 import {Member} from '../core/Member';
-import {LoggingService} from '../logging/LoggingService';
 import {ClientInfo} from '../ClientInfo';
 import HazelcastClient from '../HazelcastClient';
-import {IllegalStateError} from '../HazelcastError';
-import * as assert from 'assert';
+import {IllegalStateError, TargetDisconnectedError} from '../HazelcastError';
 import {MemberSelector} from '../core/MemberSelector';
-import {AddressHelper, DeferredPromise} from '../Util';
-import {MemberAttributeEvent, MemberAttributeOperationType} from '../core/MemberAttributeEvent';
+import {assertNotNull, DeferredPromise} from '../Util';
 import {MembershipListener} from '../core/MembershipListener';
 import {MembershipEvent} from '../core/MembershipEvent';
 import {UuidUtil} from '../util/UuidUtil';
 import {ILogger} from '../logging/ILogger';
-import Address = require('../Address');
-import ClientMessage = require('../ClientMessage');
+import {UUID} from '../core/UUID';
+import {ClientConnectionManager} from '../network/ClientConnectionManager';
+import {InitialMembershipListener} from '../core/InitialMembershipListener';
+import {InitialMembershipEvent} from '../core/InitialMembershipEvent';
+import {MemberInfo} from '../core/MemberInfo';
+import {Cluster} from '../core/Cluster';
 
 export enum MemberEvent {
     ADDED = 1,
     REMOVED = 2,
 }
 
-/**
- * Manages the relationship of this client with the cluster.
- */
-export class ClusterService {
+class MemberListSnapshot {
+    version: number;
+    members: Map<string, Member>;
```

Comment:
It might be a good idea to introduce a `readonly membersList: Member[]` field this would be initialized with `Array.from(members.values())` in the constructor. This would help to avoid litter arrays generated by `Array.from(snapshotVariable.members.values())` calls in `ClusterService`. WDYT?

Reply:
I introduced the `memberList` property but instead of initializing it in the constructor, I populate the list in the `createSnapshot` method so that we won't pass over the `members.values()` again. Is that okay ?

Reply:
Sure. This approach also works fine.